[rref] Handle exceptions returned via remote() calls

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35331 [rref] Handle exceptions returned via remote() calls**

When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

Commit:
[WIP] Handle exceptions returned via remote() calls

When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

Commit:
Update on "[WIP] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -182,23 +182,28 @@ void RequestCallbackImpl::processRpc(
+        ownerRRef->setValue(std::move(stack.front()));
+      } catch (const std::exception& e) {
+        // Don't throw in this call, but rather transfer the exception
+        // to the rref.
```
I think we should throw here to ensure that the client receives an appropriate exception (is there a reason not to?). We do attach callbacks to the Future for such messages on the client side and the error should be propagated there so that the callbacks can handle the error appropriately.

mrshenli replied:
What's the impact on he reference counting? I guess no impact, and the only difference is that the RRef is holding an error. In that case, does it mean `RemoteException` should also remove a UserRRef from `pendingUsers_`?

jjlilley replied:
This change intentionally doesn't throw here, so that the client can receive the appropriate exception.

If we throw, what happens is 
 - remoteTorchscript() call in caffe2/torch/csrc/distributed/rpc/torchscript_functions.cpp on the client side is making this call client-side
 - it runs the following:
    ctx.addPendingUser(userRRefPtr->forkId(), userRRefPtr);                                                                                                                                                          
    fm->addCallback(callback::confirmPendingUser);                                                                                                                                                                   
 - Then the confirmPendingUser call receives the exception, and blows up internally
 - The exception is never translated to the rref, and the unittest times out.

If we simply keep the flows mostly as they are, and allow the OwnerRRef to handle the exception, and propagate it to the provided future, things end up propagating as expected.

jjlilley replied:
Re: ref-counting- right, I was mostly trying to keep logic and flows as close as possible to the existing to avoid issues.

Re: pendingUsers_ - I'm trying to think, which case in particular? In this case, run the confirmPendingUser codepath just as before..

pritamdamania87 replied:
Can we update `confirmPendingUser` to not throw an exception via `handleException` and maybe just LOG the exception for now and skip executing rest of the function in case of error? The reason I feel we should throw an error here is that in general if an RPC fails, the client should be notified of this. If the client is not handling the error properly today (ex: confirmPendingUsers), we should probably fix that.

Not throwing an exception here might be fine for now, but in the future I can imagine enhancements to the RRef protocol where we might want to be notified of this error on the client side (on the appropriate Future) and maybe have some custom handling there.

To be clear, I think we should attach the error on the RRef as we do in this PR and also throw an exception.

pritamdamania87 replied:
> In that case, does it mean RemoteException should also remove a UserRRef from pendingUsers_?

Is this correct for the RRef protocol? If we've received an exception, we can't really remove the RRef from pendingUsers_ since we don't know if the owner has successfully recorded this particular fork.

mrshenli replied:
> The reason I feel we should throw an error here is that in general if an RPC fails, the client should be notified of this. 

There are multiple cases:

1. no error, remove the UserRRef from pendingUsers_
2. got exception here, remove the UserRRef from pendingUsers_  (as the fork has already been registered), and send back an exception message
3. got exception in other places (e.g., agent), send back an exception message

So for two, if we'd like to notify the client about the error, looks like we need to allow RemoteRet hold an exception message?

jjlilley replied:
So in this particular change, with the server-side changes, we won't see an exception on the client `confirmPendingUser` side in a server-side operator throws case like this.

(Things would be different if there's e.g. a network-related rpc failure, which is a separate issue... not sure if we should do something wildly different in that case)
 
Right now, if we throw from ``confirmPendingUser``, we end up throwing inside a future completion callback, inside (IIRC) a process_group_agent thread, and ends up being effectively logged and discarded, but in a somewhat unexpected manner.

We don't necessarily need to solve client side network error on remote issue in this particular change, but we do solve the issue of propagating the error to the rref.

jjlilley replied:
Shen,
- I agree that we should clean up the pendingUsers_ in a few corner cases. Thanks for pointing that out! I'll make those updates.

- If the server puts an exception in its ownerRRef, things already get propagated to the client's userRRef, and the exception is properly thrown, as the tests show. (But need to fix pendingUsers_)

- The case where there's e.g. a network error perceived by the client side, seems more complex.
   Right now, it appears that the only way to propagate back that async-generated error ``remoteTorchscript()`` is via the already-returned userRRef.
   It seems like there's no good [existing] mechanism for the client to place this observed network error into the userRRef, unless there's something I'm missing (?) Though I'm also concerned about widening the interface...
    What happens in real life now is that we simply time out the unittest, since the message never makes it back.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -225,6 +234,18 @@ void OwnerRRef::setValue(IValue&& value) {
 
+void OwnerRRef::setError(std::string error) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  error_ = error;
```
std::move here to avoid copy?

jjlilley replied:
Can't unfortunately due to the use in the callback below.
I can const ref, can update

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -261,12 +269,16 @@ void RequestCallbackImpl::processRpc(
+              try {
                 Message m = ScriptRRefFetchRet({rref->getValue()}).toMessage();
                 m.setId(messageId);
                 responseFuture->markCompleted(std::move(m));
```
Did we add these try-catch blocks here because there was something that used to throw and we weren't catching it here?

jjlilley replied:
No, we're propagating the exception in OwnerRRef now.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -190,13 +190,18 @@ RRefForkData UserRRef::fork() const {
   std::unique_lock<std::mutex> lock(mutex_);
-  valueCV_.wait(lock, [this] { return value_.has_value(); });
+  valueCV_.wait(
+      lock, [this] { return value_.has_value() || error_.has_value(); });
```
nit: can just call `hasValue` here?

jjlilley replied:
We need to break out of the loop if either there's a value or error.
There are other ways to do it, but it's a cheap check, and avoids yet-another-bool in the impl

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1844,6 +1848,14 @@ def test_rref_py_pickle_not_supported(self):
+                          raise_or_inc,
+                          args=(torch.ones(2),))
+        with self.assertRaisesRegex(Exception, ".*Expected error.*"):
+            rref.to_here()
```
Another nice test to have would be to pass this rref to another node and verify calling `rref.to_here()` throws there as well.

pritamdamania87 reviewed (changes requested):
Thanks for fixing this! Mostly requesting changes since I think we should propagate the error to the client as well and not just set the error on the RRef.

Commit:
Update on "[WIP] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

Commit:
Update on "[rref] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

Commit:
Update on "[rref] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

jjlilley commented:
I uploaded an iteration that illustrates the client propagate issue that I mentioned.

It does improves things, but in general, it seems like the error handling in rref probably needs a few more iterations.

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -204,15 +213,24 @@ PyRRef pyRemotePythonUdf(
 
-    fm->addCallback(callback::confirmPendingUser);
+    ctx.addPendingUser(userRRef->forkId(), userRRef);
+    fm->addCallback([forkId{userRRef->forkId()}](
```
This looks the same as the the other two occurrences. Any reason for not directly modify `confirmPendingUser`?

jjlilley replied:
This was a symptom of an evolving review. You're right - I just restored the ``confirmPendingUser()`` function, albeit with a new ``expectedForkId`` arg

I had to wrap it with a lambda to bind the expectedForkId arg, but looks cleaner with the suggestion.

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -157,7 +157,16 @@ PyRRef pyRemoteBuiltin(
+    fm->addCallback([forkId{userRRef->forkId()}](
+                        const rpc::Message& message,
+                        const c10::optional<utils::FutureError>& futErr) {
+      if (!futErr) {
```
I see. This is not perfect, but is better than what we currently have.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -224,15 +229,18 @@ void RequestCallbackImpl::processRpc(
+              PyObjectType::get());
+        }
+        ownerRRef->setValue(std::move(py_ivalue));
+      } catch (std::exception& e) {
```
For Python UDF, we might get `py::error_already_set` error, which requires GIL to destruct  (see #35283 )

jjlilley replied:
Thanks for the pointer! I went ahead and copied the recipe from that PR.

Commit:
Update on "[rref] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

Commit:
rebase on "[rref] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

jjlilley commented:
Any other review comments? Thanks.

Commit:
rebase on "[rref] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

Commit:
Update on "[rref] Handle exceptions returned via remote() calls"


When the function called by remote() throws, it seems sensible to
surface that exeption when rref.to_here() is called.

Doing this only involves simple modifications:
 - we need the OwnerRRef to keep around an optional<string>
   for the error
 - add an OwnerRRef setError() method that's parallel to setValue(),
   and plumb through the logic

We add rpc_tests to verify that the exception is propagated properly.

Differential Revision: [D20634078](https://our.internmc.facebook.com/intern/diff/D20634078/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -238,6 +247,18 @@ void OwnerRRef::setValue(IValue&& value) {
+  lock.unlock();
+  valueCV_.notify_all();
+  if (future.get()) {
+    future->setErrorIfNeeded(error);
```
This means we are not going to throw away the error when the future is already complete. It might need more discussion to figure out what would be the appropriate solution to handle this. For now, maybe let's log in `setErrorIfNeeded` to indicate that an error occurred after completion?

mrshenli reviewed (approved):
LGTM! Thanks! 

I created #35758 to track the followup work item.
